source/extensions/filters/network/thrift_proxy/router/router_ratelimit_impl.cc

```diff
@@ -0,0 +1,162 @@
+#include "extensions/filters/network/thrift_proxy/router/router_ratelimit_impl.h"
+
+#include "envoy/common/exception.h"
+#include "envoy/ratelimit/ratelimit.h"
+
+#include "extensions/filters/network/thrift_proxy/router/router.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace ThriftProxy {
+namespace Router {
+
+bool SourceClusterAction::populateDescriptor(const RouteEntry&, RateLimit::Descriptor& descriptor,
+                                             const std::string& local_service_cluster,
+                                             const MessageMetadata&,
+                                             const Network::Address::Instance&) const {
+  descriptor.entries_.push_back({"source_cluster", local_service_cluster});
+  return true;
+}
+
+bool DestinationClusterAction::populateDescriptor(const RouteEntry& route,
+                                                  RateLimit::Descriptor& descriptor,
+                                                  const std::string&, const MessageMetadata&,
+                                                  const Network::Address::Instance&) const {
+  descriptor.entries_.push_back({"destination_cluster", route.clusterName()});
+  return true;
+}
+
+bool RequestHeadersAction::populateDescriptor(const RouteEntry&, RateLimit::Descriptor& descriptor,
+                                              const std::string&, const MessageMetadata& metadata,
+                                              const Network::Address::Instance&) const {
+  if (use_method_name_) {
+    if (!metadata.hasMethodName()) {
+      return false;
+    }
+
+    descriptor.entries_.push_back({descriptor_key_, metadata.methodName()});
```

Comment:
I originally treated the fields that all thrift requests have in common (method name, message type, and sequence id) as fields in a simple MessageMetadata struct. In the future, there's the possibility to make HeaderMapImpl sufficiently configurable to allow the thrift proxy to specify that those fields should get special fast-path treatment (e.g. like :path) and then it would make sense to refactor the proxy to just use a HeaderMap for all these bits of data.